### PR TITLE
feat: formatOptions for number fields

### PR DESCRIPTION
**Pull request by phillipmalboeuf**
## Description

This pull request (second attempt after https://github.com/payloadcms/payload/pull/1510 was closed) offers an answer to the discussion brought up by @Rar9 and @jmikrut here: https://github.com/payloadcms/payload/discussions/1222
JavaScript's built-in [Intl.NumberFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat) offers a wide range of options from currency display, to internationalized number formats, and unit formatting. This PR only affects the dashboard, perhaps we can look into offering a way to pass the formatted value to the REST/GraphQL call, but for my use case this isn't necessary.

- [x] I have read and understand the CONTRIBUTING.md document in this repository

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Checklist:

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Existing test suite passes locally with my changes
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: src/admin/components/forms/FieldDescription/index.tsx [hunk 1]
```diff
@@ -11,13 +11,14 @@ const FieldDescription: React.FC<Props> = (props) => {
...
     value,
+    formatOptions,
   } = props;
 
   const { i18n } = useTranslation();
 
   if (isComponent(description)) {
     const Description = description;
-    return <Description value={value} />;
+    return <Description value={formatOptions ? new Intl.NumberFormat(i18n.language, formatOptions).format(value as number) : value} />;
```
Comment (jacobsfletch) [inline src/admin/components/forms/FieldDescription/index.tsx:21]: This doesn't seem right. We should not be formatting the description at all, especially considering it's accepts a callback so _you_ can format it. This also doesn't protect field types that are not numbers from running through `Intl.NumberFormat`.

File: src/admin/components/forms/field-types/Number/index.tsx [hunk 2]
```diff
@@ -86,24 +87,32 @@ const NumberField: React.FC<Props> = (props) => {
...
+          step={step}
+          onWheel={(e) => {
+            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+            // @ts-ignore
+            e.target.blur();
+          }}
+        />
+        {formatOptions && (
+          <div className="number__formatted">
+            {new Intl.NumberFormat(i18n.language, formatOptions).format(value as number)}
```
Comment (jacobsfletch) [inline src/admin/components/forms/field-types/Number/index.tsx:108]: This is great.

File: src/admin/components/views/collections/List/Cell/index.tsx [hunk 3]
```diff
@@ -56,7 +57,7 @@ const DefaultCell: React.FC<Props> = (props) => {
       <WrapElement {...wrapElementProps}>
         {((cellData === '' || typeof cellData === 'undefined') && 'label' in field) && t('noLabel', { label: getTranslation(typeof field.label === 'function' ? 'data' : field.label || 'data', i18n) })}
         {typeof cellData === 'string' && cellData}
-        {typeof cellData === 'number' && cellData}
+        {typeof cellData === 'number' && ((field as NumberField).admin.formatOptions ? new Intl.NumberFormat(i18n.language, (field as NumberField).admin.formatOptions).format(cellData) : cellData)}
```
Comment (jacobsfletch) [inline src/admin/components/views/collections/List/Cell/index.tsx:60]: This too.

**Review by jacobsfletch — CHANGES_REQUESTED** [review]
@phillipmalboeuf just a couple small changes and we could probably get this merged in 🙌. Nice work!

File: src/admin/components/forms/field-types/Number/index.tsx (hunk 2 above)
Comment (jacobsfletch) [reply src/admin/components/forms/field-types/Number/index.tsx:108]: Err, we need to format display `value` on line 93 actually—not render an additional div next to the input. I didn't catch this at first pass.

File: src/admin/components/forms/FieldDescription/index.tsx (hunk 1 above)
Comment (phillipmalboeuf) [reply src/admin/components/forms/FieldDescription/index.tsx:21]: I disagree here @jacobsfletch, this is not formatting the description, only the value passed to it. And at the moment, the description functions or components aren't being passed the `i18n.language`. If the formatted values aren't passed to the Description, the language would need to be passed to it at the very least.

And to your second comment, only number fields have `formatOptions` at the moment.

File: src/admin/components/forms/field-types/Number/index.tsx (hunk 2 above)
Comment (phillipmalboeuf) [reply src/admin/components/forms/field-types/Number/index.tsx:108]: This is not possible, the input type is set to `number`. For what you are asking, I believe we would need to set it to `type='tel'`, and to include some new javascript package that auto-formats the input as the user is typing. I disagree that we need to format inline on the input.

File: src/admin/components/forms/FieldDescription/index.tsx (hunk 1 above)
Comment (jacobsfletch) [reply src/admin/components/forms/FieldDescription/index.tsx:21]: Passing the language through the description callback sounds fine to me. Then back to that second point, you are typing the value `as number` which clearly illustrates the problem.

File: src/admin/components/forms/field-types/Number/index.tsx (hunk 2 above)
Comment (jacobsfletch) [reply src/admin/components/forms/field-types/Number/index.tsx:108]: The ultimate goal here is to format only the display value of the field without saving to the db, exactly as we currently do for dates. I see your point about the input being of `type="number"`. The solution here is to absolutely position a text field overtop the number field, and format the value of _that_. Then sync the value of the number field beneath, which is hidden.

Comment (phillipmalboeuf) [reply src/admin/components/forms/field-types/Number/index.tsx:108]: The date field relies on `react-datepicker`, I have attempted to use the `react-number-format` library but it doesn't handle decimal values with Intl.NumberFormat, not sure why.

File: src/admin/components/forms/FieldDescription/index.tsx (hunk 1 above)
Comment (phillipmalboeuf) [reply src/admin/components/forms/FieldDescription/index.tsx]: The latest commits have pulled formatting out of the Description, and added the `language` argument to the description function.

**Comment by jacobsfletch** [comment]
@phillipmalboeuf I've been thinking about this one quite a bit and I do think there's a ton of value in internationalizing our number fields 👍. I want to quickly summarize some of the requirements here and what needs to be done yet (some of this is already done):
1. The _value_ itself must remain completely unchanged, it must save to the db as a raw number exactly as it does now
2. We need to format the _appearance_ of the number by hiding the true input seamlessly behind a formatted text field
3. The formatting needs to be done via [Intl.NumberFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat), this means an options object must be exposed in the config

The second item in that list is the undoubtedly the hardest part because the formatter adds commas, decimals, prefixes, suffixes, etc in an unpredictable way. Although we can easily format _from_ a number, there is no API to decode an already formatted number. We need to find a good, stable way to do this and it must appear completely natural to the user.

After a quick search on I found this nice article from Adobe's "React Spectrum" design system talking about _this exact thing_ 🙌  : https://react-spectrum.adobe.com/blog/how-we-internationalized-our-numberfield.html

**Comment by phillipmalboeuf** [comment]
Sounds good @jacobsfletch! So I got very close with [last commit](https://github.com/payloadcms/payload/pull/1924/commits/72ea19653a601ab2628e5ed19d461bae53bdb7c5) using @internationalized/number (good find btw!), but there's some input cursor jumps occurring. Not sure how to address, any ideas?

**Comment by PatrikKozak** [comment]
Hey @phillipmalboeuf - Are you still in need of any assistance here?

**Comment by denolfe** [comment]
Hey @phillipmalboeuf, this would still be a great addition to Payload. We've recently migrated our repository to a monorepo, which means in order for this PR to be viable, you'll need to update it to target the `main` branch and change the code to target `packages/payload`.

Let me know if this is something you can take on.

**Pull request closed without merging**